File: week3/ReverseLinkedList.py
//Truc's solution: 

          // Recrusion: 
          class Solution(object):
              def reverseList(self, head):
                  if not head or not head.next: 
                      return head

                  result = self.reverseList(head.next)
                  head.next.next = head
                  head.next = None

                  return result

              // Runtime: O(n), space: O(n)

          // Iteration:
          class Solution(object):
              def reverseList(self, head):
                  if not head: 
                      return None
                  prev = None
                  cur = head
                  while cur: 
                      nex = cur.next
                      cur.next = prev
                      prev = cur
                      cur = nex
                  return prev
               // Runtime: O(n), space: O(1)

   
  

File: week3/ReverseLinkedList.java
// Hue's solution: 

class Solution {
    public ListNode reverseList(ListNode head) {
        if(head == null){
            return null;
        }
        return reverse(head).get(0);
    }
    //check head is the end
    //create new linked list
    //return a list of head and end of a linked list
    private ArrayList<ListNode> reverse(ListNode head){
        if(head.next == null){
            ListNode newList = new ListNode(head.val);
            ArrayList<ListNode> result = new ArrayList();
            result.add(newList);
            result.add(newList);
            return result;
        }
        ArrayList<ListNode> reversed = reverse(head.next);
        System.out.println(reversed.toString());
        reversed.get(1).next = new ListNode(head.val);
        reversed.set(1, reversed.get(1).next);
        return reversed;
    }
}
